File: employeeImportance/employeeImportance.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Employee {
public:
    Employee(int _id, int _importance) : id(_id), importance(_importance) {
    
    }
    int id;
    int importance;
    vector<int> subordinates;
};

class Solution {
public:
    int getImportance(vector<Employee*> employees, int id) {
        if (employees.size() == 0)
            return 0;
        unordered_map<int, Employee*> m;
        for (Employee* e : employees) {
            m[e->id] = e;
        }
        return dfs(m, id);
    }
    int dfs(unordered_map<int, Employee*>& m, int id) {
        int res = m[id]->importance;
        for (int i = 0; i < m[id]->subordinates.size(); i++) {
            res += dfs(m, m[id]->subordinates[i]);       
        }
        return res;
    }
};

int main()
{
    Solution s;
    Employee n0(1,5), n1(2,3), n2(3,3);
    n0.subordinates.push_back(n1.id);
    n0.subordinates.push_back(n2.id);
    vector<Employee*> employees;
    employees.push_back(&n0);
    employees.push_back(&n1);
    employees.push_back(&n2);
    cout << s.getImportance(employees, n0.id) << endl;
    return 0;
}

File: 690.employeeImportance/employeeImportance.java
/*
// Employee info
class Employee {
    // It's the unique id of each node;
    // unique id of this employee
    public int id;
    // the importance value of this employee
    public int importance;
    // the id of direct subordinates
    public List<Integer> subordinates;
};
*/
class Solution {
    public int getImportance(List<Employee> employees, int id) {
        if (employees == null || employees.size() == 0)
            return 0;
        Map<Integer, Employee> m = new HashMap<>();
        for (Employee e : employees)
            m.put(e.id, e);
        return dfs(m, id);
    }
    int dfs(Map<Integer, Employee> m, int id) {
        int res = m.get(id).importance;
        for (int sub : m.get(id).subordinates)
            res += dfs(m, sub);
        return res;
    }
}
